File: src/multivariable_calculus_directional_derivatives_medium.py
"""
multivariable_calculus - directional_derivatives (medium)
Generated: 2026-02-22T05:36:28.573734
"""

from problem_utils import *

def generate():
    # Pick a difficulty level within the standard range
    difficulty_level = randint(1300, 1600)
    
    if difficulty_level < 1400:
        # Easier: directional derivative with simple function and standard unit vector
        # Example: f(x,y) = x^2 + y^2, direction vector <1,0> or <0,1>
        func_type = randint(1, 2)
        
        if func_type == 1:
            # Quadratic polynomial
            a_val = nonzero(-3, 3)
            b_val = nonzero(-3, 3)
            f = a_val * x**2 + b_val * y**2
        else:
            # Linear combination
            a_val = nonzero(-4, 4)
            b_val = nonzero(-4, 4)
            f = a_val * x + b_val * y
        
        # Use a coordinate direction
        direction_choice = randint(1, 2)
        if direction_choice == 1:
            u_vec = [1, 0]
            u_latex = "\\langle 1, 0 \\rangle"
        else:
            u_vec = [0, 1]
            u_latex = "\\langle 0, 1 \\rangle"
        
        # Point of evaluation
        x0, y0 = randint(-3, 3), randint(-3, 3)
        
        # Compute gradient
        fx = diff(f, x)
        fy = diff(f, y)
        
        # Evaluate gradient at point
        fx_val = fx.subs([(x, x0), (y, y0)])
        fy_val = fy.subs([(x, x0), (y, y0)])
        
        # Directional derivative = grad f · u
        ans = fx_val * u_vec[0] + fy_val * u_vec[1]
        
        return problem(
            question=f"Find the directional derivative of $f(x,y) = {latex(f)}$ at the point $({x0}, {y0})$ in the direction ${u_latex}$.",
            answer=ans,
            difficulty=difficulty_level,
            topic="multivariable_calculus/directional_derivatives",
            solution=steps(
                f"Compute the gradient: $\\nabla f = \\langle {latex(fx)}, {latex(fy)} \\rangle$",
                f"Evaluate at $({x0}, {y0})$: $\\nabla f({x0}, {y0}) = \\langle {fx_val}, {fy_val} \\rangle$",
                f"The directional derivative is $D_{{\\mathbf{{u}}}} f = \\nabla f \\cdot \\mathbf{{u}} = {fx_val} \\cdot {u_vec[0]} + {fy_val} \\cdot {u_vec[1]} = {ans}$"
            ),
            calculator="scientific"
        )
    
    else:
        # Harder: non-unit direction vector requiring normalization
        func_type = randint(1, 3)
        
        if func_type == 1:
            # Quadratic
            a_val = nonzero(-3, 3)
            b_val = nonzero(-3, 3)
            c_val = nonzero(-3, 3)
            f = a_val * x**2 + b_val * x * y + c_val * y**2
        elif func_type == 2:
            # Polynomial
            a_val = nonzero(-3, 3)
            b_val = nonzero(-3, 3)
            f = a_val * x**2 + b_val * y**2
        else:
            # Mixed
            a_val = nonzero(-2, 2)
            b_val = nonzero(-2, 2)
            f = a_val * x * y + b_val * (x**2 + y**2)
        
        # Non-unit direction vector
        v1, v2 = nonzero(-4, 4), nonzero(-4, 4)
        v_latex = f"\\langle {v1}, {v2} \\rangle"
        
        # Normalize
        v_mag = sqrt(v1**2 + v2**2)
        u1 = v1 / v_mag
        u2 = v2 / v_mag
        
        # Point of evaluation
        x0, y0 = randint(-2, 2), randint(-2, 2)
        
        # Compute gradient
        fx = diff(f, x)
        fy = diff(f, y)
        
        # Evaluate gradient at point
        fx_val = fx.subs([(x, x0), (y, y0)])
        fy_val = fy.subs([(x, x0), (y, y0)])
        
        # Directional derivative = grad f · u
        ans = simplify(fx_val * u1 + fy_val * u2)
        
        return problem(
            question=f"Find the directional derivative of $f(x,y) = {latex(f)}$ at the point $({x0}, {y0})$ in the direction of the vector ${v_latex}$.",
            answer=ans,
            difficulty=difficulty_level,
            topic="multivariable_calculus/directional_derivatives",
            solution=steps(
                f"First, normalize the direction vector: $\\mathbf{{u}} = \\frac{{{v_latex}}}{{\\|{v_latex}\\|}} = \\frac{{{v_latex}}}{{{latex(v_mag)}}}$",
                f"Compute the gradient: $\\nabla f = \\langle {latex(fx)}, {latex(fy)} \\rangle$",
                f"Evaluate at $({x0}, {y0})$: $\\nabla f({x0}, {y0}) = \\langle {fx_val}, {fy_val} \\rangle$",
                f"The directional derivative is $D_{{\\mathbf{{u}}}} f = \\nabla f \\cdot \\mathbf{{u}} = \\frac{{{fx_val} \\cdot {v1} + {fy_val} \\cdot {v2}}}{{{latex(v_mag)}}} = {latex(ans)}$"
            ),
            calculator="scientific"
        )

emit(generate())

File: src/multivariable_calculus_directional_derivatives_medium.jl
# multivariable_calculus - directional_derivatives (medium)
# Generated: 2026-03-08T20:59:38.301181

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y z begin
    set_topic!("multivariable_calculus/directional_derivatives")
    
    problem_type = rand(1:4)
    
    if problem_type == 1
        # Compute directional derivative at a point given f and direction vector
        # ELO: 1200-1400
        a = randint(-5, 5)
        b = randint(-5, 5)
        c = randint(-5, 5)
        f = a*x^2 + b*y^2 + c*x*y
        
        x0 = randint(-4, 4)
        y0 = randint(-4, 4)
        
        # Direction vector
        v1 = nonzero(-5, 5)
        v2 = nonzero(-5, 5)
        mag = sqrt(v1^2 + v2^2)
        
        # Gradient at point
        fx = diff(f, x)
        fy = diff(f, y)
        fx_val = substitute(fx, Dict(x => x0, y => y0))
        fy_val = substitute(fy, Dict(x => x0, y => y0))
        
        # Directional derivative
        ans = (fx_val * v1 + fy_val * v2) / mag
        
        problem(
            question="Let \$f(x,y) = $(tex(f))\$. Find the directional derivative of \$f\$ at the point \$($(x0), $(y0))\$ in the direction of the vector \$\\mathbf{v} = \\langle $(v1), $(v2) \\rangle\$.",
            answer=ans,
            difficulty=(1200, 1400),
            solution=steps(
                sol("Given function", f),
                "Compute gradient: \$\\nabla f = \\langle $(tex(fx)), $(tex(fy)) \\rangle\$",
                "At point \$($(x0), $(y0))\$: \$\\nabla f($(x0), $(y0)) = \\langle $(fx_val), $(fy_val) \\rangle\$",
                "Unit vector: \$\\mathbf{u} = \\frac{1}{$(tex(mag))} \\langle $(v1), $(v2) \\rangle\$",
                "Directional derivative: \$D_{\\mathbf{u}}f = \\nabla f \\cdot \\mathbf{u} = \\frac{$(fx_val)($(v1)) + $(fy_val)($(v2))}{$(tex(mag))}\$",
                sol("Answer", ans)
            ),
            time=120
        )
        
    elseif problem_type == 2
        # Find direction of maximum rate of increase
        # ELO: 1300-1500
        a = nonzero(-4, 4)
        b = nonzero(-4, 4)
        c = randint(-3, 3)
        f = a*x^2 + b*x*y + c*y
        
        x0 = randint(-3, 3)
        y0 = randint(-3, 3)
        
        fx = diff(f, x)
        fy = diff(f, y)
        fx_val = substitute(fx, Dict(x => x0, y => y0))
        fy_val = substitute(fy, Dict(x => x0, y => y0))
        
        ans_str = "\\langle $(fx_val), $(fy_val) \\rangle"
        
        problem(
            question="Let \$f(x,y) = $(tex(f))\$. Find the direction (as a vector) in which \$f\$ increases most rapidly at the point \$($(x0), $(y0))\$.",
            answer=ans_str,
            difficulty=(1300, 1500),
            answer_type="vector",
            solution=steps(
                sol("Given function", f),
                "The direction of maximum increase is the gradient vector",
                "Compute: \$\\frac{\\partial f}{\\partial x} = $(tex(fx))\$, \$\\frac{\\partial f}{\\partial y} = $(tex(fy))\$",
                "At point \$($(x0), $(y0))\$: \$\\nabla f($(x0), $(y0)) = \\langle $(fx_val), $(fy_val) \\rangle\$",
                sol("Answer", ans_str)
            ),
            time=90
        )
        
    elseif problem_type == 3
        # Compute maximum rate of change (magnitude of gradient)
        # ELO: 1400-1600
        a = nonzero(-3, 3)
        b = nonzero(-3, 3)
        c = nonzero(-3, 3)
        f = a*x*y + b*x + c*y
        
        x0 = randint(-4, 4)
        y0 = randint(-4, 4)
        
        fx = diff(f, x)
        fy = diff(f, y)
        fx_val = substitute(fx, Dict(x => x0, y => y0))
        fy_val = substitute(fy, Dict(x => x0, y => y0))
        
        ans = sqrt(fx_val^2 + fy_val^2)
        
        problem(
            question="Let \$f(x,y) = $(tex(f))\$. Find the maximum rate of change of \$f\$ at the point \$($(x0), $(y0))\$.",
            answer=ans,
            difficulty=(1400, 1600),
            solution=steps(
                sol("Given function", f),
                "Maximum rate of change equals \$\\|\\nabla f\\|\$",
                "Compute gradient: \$\\nabla f = \\langle $(tex(fx)), $(tex(fy)) \\rangle\$",
                "At point \$($(x0), $(y0))\$: \$\\nabla f = \\langle $(fx_val), $(fy_val) \\rangle\$",
                "Magnitude: \$\\|\\nabla f\\| = \\sqrt{($(fx_val))^2 + ($(fy_val))^2}\$",
                sol("Answer", ans)
            ),
            time=120
        )
        
    else
        # Directional derivative equals given value - find relationship
        # ELO: 1500-1800
        a = nonzero(-4, 4)
        b = nonzero(-4, 4)
        f = a*x^2 + b*y^2
        
        x0 = nonzero(-3, 3)
        y0 = nonzero(-3, 3)
        
        fx = diff(f, x)
        fy = diff(f, y)
        fx_val = substitute(fx, Dict(x => x0, y => y0))
        fy_val = substitute(fy, Dict(x => x0, y => y0))
        
        # Ask for directional derivative in direction making angle θ with positive x-axis
        target_val = randint(-20, 20)
        
        # D_u f = fx_val * cos(θ) + fy_val * sin(θ) = target_val
        # This is for a specific angle, but we'll ask them to compute it
        
        theta_deg = choice([0, 30, 45, 60, 90, 120, 135, 150, 180])
        theta_rad = theta_deg * π / 180
        
        u1 = cos(theta_rad)
        u2 = sin(theta_rad)
        
        ans = fx_val * u1 + fy_val * u2
        
        problem(
            question="Let \$f(x,y) = $(tex(f))\$. Find the directional derivative of \$f\$ at the point \$($(x0), $(y0))\$ in the direction making an angle of \$$(theta_deg)^\\circ\$ with the positive \$x\$-axis.",
            answer=ans,
            difficulty=(1500, 1800),
            solution=steps(
                sol("Given function", f),
                "Direction vector: \$\\mathbf{u} = \\langle \\cos($(theta_deg)^\\circ), \\sin($(theta_deg)^\\circ) \\rangle = \\langle $(tex(u1)), $(tex(u2)) \\rangle\$",
                "Compute gradient: \$\\nabla f = \\langle $(tex(fx)), $(tex(fy)) \\rangle\$",
                "At \$($(x0), $(y0))\$: \$\\nabla f = \\langle $(fx_val), $(fy_val) \\rangle\$",
                "Directional derivative: \$D_{\\mathbf{u}}f = \\nabla f \\cdot \\mathbf{u} = $(fx_val) \\cdot $(tex(u1)) + $(fy_val) \\cdot $(tex(u2))\$",
                sol("Answer", ans)
            ),
            time=150
        )
    end
end